core/App/ehsm_napi.cpp

```diff
@@ -462,26 +573,69 @@ char* NAPI_Sign(const char* cmk_base64,
         const char* digest)
 {    
     RetJsonObj retJsonObj;
-    string cmk_str;
-    ehsm_keyblob_t masterkey;
     ehsm_status_t ret = EH_OK;
-    ehsm_data_t digest_data;
-    ehsm_data_t signature;
+    ehsm_keyblob_t cmk = {0};
+    ehsm_data_t digest_data = {0};
+    ehsm_data_t signature = {0};
+
+    string cmk_str;
     string signature_base64;
 
+    int cmk_len;
+    int digest_len;
+    int digest_maxLen = 0;
+
     cmk_str = base64_decode(cmk_base64);
-    ret = ehsm_deserialize_cmk(&masterkey, (const uint8_t*)cmk_str.data(), cmk_str.size());
+    cmk_len = cmk_str.size();
+    digest_len = strlen(digest);
+
+    if(cmk_len == 0 || cmk_len > EH_CMK_MAX_SIZE){
+        retJsonObj.setCode(retJsonObj.CODE_BAD_REQUEST);
+        retJsonObj.setMessage("The cmk's length is invalid.");
+        goto out;
+    }
+
+    ret = ehsm_deserialize_cmk(&cmk, (const uint8_t*)cmk_str.data(), cmk_len);
     if (ret != EH_OK) {
         retJsonObj.setCode(retJsonObj.CODE_FAILED);
         retJsonObj.setMessage("Server exception.");
         goto out;
     }
 
-    digest_data.datalen = strlen(digest);
+    switch (cmk.metadata.keyspec)
+    {
+        case EH_RSA_2048:
+            // TODO
+            break;
+        case EH_RSA_3072:
+            digest_maxLen = RSA_OAEP_3072_DIGEST_SIZE;
+            break;
+        case EH_EC_P256:
+            // TODO
+            break;
+        case EH_EC_P512:
+            // TODO
+            break;
+        case EH_EC_SM2:
+            // TODO
+            break;
+        default:
+            retJsonObj.setCode(retJsonObj.CODE_BAD_REQUEST);
+            retJsonObj.setMessage("The cmk's keyspec is invalid.");
+            goto out;
+    }
+
```

Comment:
don't think you need the switch() from line#605-line#627, because the sign operation only accept the parameter as digest which is generated for the original message by using a hash algorithm. The hash algorithm is specified by the Algorithm parameter.

in most cases the digest is 256 bytes. or not longer than it.

Reply:
It will be fix.